### PR TITLE
Raise TypeError on assigned grad with wrong type

### LINKED ISSUE #64813: Segfault on setting gradient value to instance of user-defined class
## 🐛 Bug

Seems to be a segfault when setting the value of a parameter.grad to a user-defined class instance. This is probably not useful, but I wanted to document it. It would be better for this crash to raise an error rather than issue a segfault if possible.

## To Reproduce

Minimal reproduction code:

```
class Example:
    def __init__(self):
        self.x = 0
        
x = torch.tensor([1., 2., 3., 4., 5.])
model = torch.nn.Linear(5, 1)
model(x).backward()

model.weight.grad = Example()
```

## Expected behavior

I imagine there could be some expected behavior, but this should probably remain undefined, but should be gracefully handled by raising an exception rather than a segfault.

## Environment

PyTorch Version: 1.10.0.dev20210831

## Additional context

I came across this during a workflow where I set `param.grad = crypten.cryptensor(param.grad)` and manipulate it in CrypTen. Instead, I have to use a new variable then decrypt before setting value into `param.grad`. This is fine, however during testing, my multiprocess tests did not catch the issue even with this failure due to the segfault. Had this raised an exception, my test would have recognized that this shouldn't pass. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by Varal7**
Fixes #64813

Raises a TypeError when assigned value to a grad is not a Tensor or
None.

Adds tests.

BC-breaking note:
Setting the `.grad` field with a non-None and non-Tensor object used to return a `RuntimeError` but it now properly returns a `TypeError`.
If your code was catching this error, you should simply update it to catch the new error type.

cc @ezyang @gchanan

### CODE REVIEW

**Review by soulitzer — DISMISSED** [review]
Nice fix!

**Comment by Varal7** [comment]
So uh. Looks like Edward did remove the check on purpose. 

image

This test needs to be addresses before we carry on with the fix

**Comment by soulitzer** [comment]
Hmm from the test failure - looks like we should allow for tensor subclasses as well?

**Comment by Varal7** [comment]
I don't know if @ezyang meant to use `torch._C._TensorBase` instead of `torch.Tensor` for a specific reason, but subclasses of `torch.Tensor` work fine with `THPVariable_Check`.

**Review by anjali411 — APPROVED** [review]

**Review by soulitzer — APPROVED** [review]

**Comment by ezyang** [comment]
@albanD and I were chatting about this and it's not entirely clear you want to force grad to have exactly the same type as the base tensor. For example, when doing sparse gradients, grad will be a sparse tensor and then get densified when the optimizer applies it to the parameter.

**Comment by albanD** [comment]
Also @ezyang is subclassing `_TensorBase` directly something we want to do ever? That would mean you don't get all the logic we add in the python side (including the default `__torch_function__`, serialization, deepcopy, etc).

**Comment by Varal7** [comment]
> @albanD and I were chatting about this and it's not entirely clear you want to force grad to have exactly the same type as the base tensor. 

I don't think we're enforcing this currently or with this PR?
The following works:

```python
class MyTensor(torch.Tensor):
    pass

x = torch.ones(5)

x.grad = MyTensor(torch.ones(5))
```

**Review by albanD — APPROVED** [review]
LGTM

Note that this is BC-breaking from 1.9 as we change a RuntimeError into a TypeError (but that's an ok BC-breaking).

**Comment by ezyang** [comment]
> Also @ezyang is subclassing _TensorBase directly something we want to do ever? That would mean you don't get all the logic we add in the python side (including the default __torch_function__, serialization, deepcopy, etc).

I used to think maybe this should be allowed (in case you didn't want any of the Python methods defined in Tenosr itself), but I've changed my mind, mostly because it's simpler not to allow it. We may need to expunge some occurrences of this pattern.

> I don't think we're enforcing this currently or with this PR?

Ooops, my bad! This PR is good, no problems from me.

**Review by ezyang — APPROVED** [review]
